cfmm.mligo

```diff
@@ -293,6 +302,76 @@ let cash_transfer (storage : storage) (from : address) (to_ : address) (cash_amo
 #endif
 #endif
 
+(* Difference Equations *)
+// The Isoutility Function in https://hackmd.io/MkPSYXDsTf-giBprcDrc3w  
+(*
+// b is 2^48 below, implemented as bitwise shifts
+let isoutility (target, cash, token : nat * nat * nat) : nat = 
+    let x = cash in 
+    let y = token in 
+    let a = target in 
+    let a2 = a * a in
+    let ax2 = a2 * x * x in 
+    let by2 = Bitwise.shift_left (y * y) 96n in
+    (Bitwise.shift_right (abs((a * x * y) * (ax2 + by2) / (2 * a2))) 48n)
+
+// Returns the price dy/dx of the isoutility function, i.e. a map by multiplication ∆x => ∆y, at a given point (x,y)
+let price_cash_to_token (target : nat) (cash : nat) (token : nat) : nat = 
+    let (x,y) = (cash, token) in
+    let a = target in 
+    let ax2 = x * x * a * a in
+    let by2 = Bitwise.shift_left (y * y) 96n in
+    let num = y * (3n * ax2 + by2) in
+    let denom = x * (ax2 + 3n * by2) in
+    num/denom
+*)
+// A function to transfer assets while maintaining a constant isoutility
+let rec newton_dx_to_dy (x, y, dx, dy_approx, target, rounds : nat * nat * nat * nat * nat * int) : nat = 
+    if (rounds <= 0) (* Newton generally converges in 4 rounds, so we bound computation there *)
+    then 
+        dy_approx
+    else 
+        let a = target in 
+        let xp = x + dx in
+        let yp = y - dy_approx in 
+        let ax2 = a * a * x * x in let by2 = Bitwise.shift_left (y * y) 96n in // (y * y) * b * b
+        let axp2 = a * a * xp * xp in let byp2 = Bitwise.shift_left (abs(yp * yp)) 96n in // (abs(yp * yp)) * b * b
+        (* Newton descent formula *)
+        let num = abs (xp * yp * (axp2 + byp2) - x * y * (ax2 + by2)) in // num is always positive, even without abs
+        let denom = xp * (axp2 + 3 * byp2) in
+        let adjust = num / denom in
+        let new_dy_approx = abs (dy_approx + adjust) in
+        let next_round = (rounds - 1) in
+        newton_dx_to_dy (x,y,dx,new_dy_approx,target,next_round)
+    (*
+        if denom = 0, then either:
+        1. xp = 0 => x + dx = 0, which we don't allow, or
+        2. a*xp = 0 and b*yp = 0 => a = 0 and (b = 0 or yp = 0), which implies
+           that the price target is 0.
+     *)
+
+// A function that outputs dy (diff_token) given x, y, and dx
+let trade_dcash_for_dtoken (x : nat) (y : nat) (dx : nat) (target : nat) (rounds : int) : nat = 
+    let dy_approx = 0n in // start at 0n to always be an underestimate
+
+    if (y - dy_approx <= 0)
+    then
+        (failwith error_TOKEN_POOL_MINUS_TOKENS_WITHDRAWN_IS_NEGATIVE : nat)
+    else 
+        newton_dx_to_dy (x, y, dx, dy_approx, target, rounds)
+
+// A function that outputs dx (diff_cash) given target, x, y, and dy
+let trade_dtoken_for_dcash (x : nat) (y : nat) (dy : nat) (target : nat) (rounds : int) : nat = 
+    let dx_approx = 0n in // start at 0n to always be an underestimate
+    let target_inv = (Bitwise.shift_left 1n 96n) / target in // b^2 / a
```

Comment:
don't compute target_inv as this involves a loss of precision, pass a fraction to newton

Reply:
I just pushed a modified Newton calculation that takes the target parameter as `(a,b)` rather than just `a`. That's what you mean, correct?